lib/src/util/side_effect_visitor.dart

```diff
@@ -0,0 +1,546 @@
+  }
+
+  @override
+  visitImportDirective(ImportDirective node) {
+    super.visitImportDirective(node);
+  }
+
+  @override
+  visitIndexExpression(IndexExpression node) {
+    _hasSideEffect = true;
+  }
+
+  @override
+  visitInstanceCreationExpression(InstanceCreationExpression node) {
+    _hasSideEffect = true;
+  }
+
+  @override
+  visitIntegerLiteral(IntegerLiteral node) {
+    super.visitIntegerLiteral(node);
+  }
+
+  @override
+  visitInterpolationExpression(InterpolationExpression node) {
+    super.visitInterpolationExpression(node);
+  }
+
+  @override
+  visitInterpolationString(InterpolationString node) {
+    super.visitInterpolationString(node);
+  }
+
+  @override
+  visitIsExpression(IsExpression node) {
+    super.visitIsExpression(node);
+  }
+
+  @override
+  visitLabel(Label node) {
+    super.visitLabel(node);
+  }
+
+  @override
+  visitLabeledStatement(LabeledStatement node) {
+    super.visitLabeledStatement(node);
+  }
+
+  @override
+  visitLibraryDirective(LibraryDirective node) {
+    super.visitLibraryDirective(node);
+  }
+
+  @override
+  visitLibraryIdentifier(LibraryIdentifier node) {
+    super.visitLibraryIdentifier(node);
+  }
+
+  @override
+  visitListLiteral(ListLiteral node) {
+    super.visitListLiteral(node);
+  }
+
+  @override
+  visitMapLiteral(MapLiteral node) {
+    super.visitMapLiteral(node);
+  }
+
+  @override
+  visitMapLiteralEntry(MapLiteralEntry node) {
+    super.visitMapLiteralEntry(node);
+  }
+
+  @override
+  visitMethodDeclaration(MethodDeclaration node) {
+    super.visitMethodDeclaration(node);
+  }
+
+  @override
+  visitMethodInvocation(MethodInvocation node) {
+    _hasSideEffect = true;
+  }
+
+  @override
+  visitNamedExpression(NamedExpression node) {
+    super.visitNamedExpression(node);
+  }
+
+  @override
+  visitNativeClause(NativeClause node) {
+    super.visitNativeClause(node);
+  }
+
+  @override
+  visitNativeFunctionBody(NativeFunctionBody node) {
+    super.visitNativeFunctionBody(node);
+  }
+
+  @override
+  visitNullLiteral(NullLiteral node) {
+    super.visitNullLiteral(node);
+  }
+
+  @override
+  visitParenthesizedExpression(ParenthesizedExpression node) {
+    super.visitParenthesizedExpression(node);
+  }
+
+  @override
+  visitPartDirective(PartDirective node) {
+    super.visitPartDirective(node);
+  }
+
+  @override
+  visitPartOfDirective(PartOfDirective node) {
+    super.visitPartOfDirective(node);
+  }
+
+  @override
+  visitPostfixExpression(PostfixExpression node) {
+    _hasSideEffect = true;
+  }
+
+  @override
+  visitPrefixedIdentifier(PrefixedIdentifier node) {
+    super.visitPrefixedIdentifier(node);
+  }
+
+  @override
+  visitPrefixExpression(PrefixExpression node) {
+    _hasSideEffect = true;
+  }
+
+  @override
+  visitPropertyAccess(PropertyAccess node) {
+    super.visitPropertyAccess(node);
+  }
+
+  @override
+  visitRedirectingConstructorInvocation(RedirectingConstructorInvocation node) {
+    _hasSideEffect = true;
```

Comment:
I don't think you should ever visit one of these, but it isn't clear to me whether this should be considered to have a side-effect. Depending on your reasoning, visitSuperConstructorInvocation might want to be consistent. (But then there's always a super constructor invocation even when it isn't explicit.)